torch/autograd/__init__.py

```diff
@@ -21,21 +21,35 @@
 from . import functional
 from . import forward_ad
 from . import graph
+from .. import _vmap_internals
 
 __all__ = ['Variable', 'Function', 'backward', 'grad_mode']
 
 _OptionalTensor = Optional[torch.Tensor]
 
-def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor]) -> Tuple[_OptionalTensor, ...]:
+def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor],
+                is_grads_batched: bool) -> Tuple[_OptionalTensor, ...]:
     new_grads: List[_OptionalTensor] = []
     for out, grad in zip(outputs, grads):
         if isinstance(grad, torch.Tensor):
-            if not out.shape == grad.shape:
-                raise RuntimeError("Mismatch in shape: grad_output["
-                                   + str(grads.index(grad)) + "] has a shape of "
-                                   + str(grad.shape) + " and output["
-                                   + str(outputs.index(out)) + "] has a shape of "
-                                   + str(out.shape) + ".")
+            grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
+            if not out.shape == grad_shape:
+                if is_grads_batched:
+                    raise RuntimeError("If `is_grads_batched=True`, we interpret the first "
+                                       "dimension of each grad_output as the batch dimension. "
+                                       "The sizes of the remaining dimensions are expected to match "
+                                       "the shape of corresponding output, but a mismatch "
```

Comment:
Maybe suggest using vmap if the user actually wanted only some of their grad_output to be batched